### PR TITLE
Implement modularization of aclint

### PR DESCRIPTION
**Pull request by PorterLu**
**Related issue**: Implement modularization of aclint, which divide clint into mtimer and mswi.

**Type of change**: feature request

**Impact**: no functional change

**Development Phase**: implementation

**Release Notes**

This PR implements the modular features in ACLINT, dividing CLINT into MTIMER and MSWI. Users can use CLINTKey to generate an ACLINT. When isACLINT is false, ACLINT will be compatible with CLINT, otherwise it will not be compatible.

### CODE REVIEW

**Comment by jerryz123** [comment]
Can you summarize the changes and the justification in this PR thread?

**Comment by ZenithalHourlyRate** [comment]
> Can you summarize the changes and the justification in this PR thread?

ACLINT is an "advanced" version of CLINT and can be made compatible with CLINT, and it has spec (though in draft state, and not updated for a year).

I've had some prior communication with the PR author. I intend to replace CLINT with ACLINT in all the rocket-chip codebase, and only leave `CLINT.scala` there for other projects using it. 

For end users, that might be a breaking change as the device tree will change. We may mitigate this issue by adding a compatible string `riscv,clint0` for aclint when the base addresses are placed the same as CLINT. 

> Is it possible to add ACLINT as a separate peripheral device, and allow user to select between CLINT and ACLINT in the Config?

Are there cases where people really need that specific CLINT RTL instead of a compatible ACLINT?

Another benefit of this is that SSWI can be implemented and will be implemented in another PR.

**Comment by jerryz123** [comment]
> > Can you summarize the changes and the justification in this PR thread?
> 
> ACLINT is an "advanced" version of CLINT and can be made compatible with CLINT, and it has spec (though in draft state, and not updated for a year).

Is the spec frozen? How is software support for ACLINT?

> I've had some prior communication with the PR author. I intend to replace CLINT with ACLINT in all the rocket-chip codebase, and only leave `CLINT.scala` there for other projects using it.

What is the justification for replacing CLINT with ACLINT in default designs? 

> For end users, that might be a breaking change as the device tree will change. We may mitigate this issue by adding a compatible string `riscv,clint0` for aclint when the base addresses are placed the same as CLINT.

I prefer that we allow both a) using existing CLINT, with no changes to DTS, or b) using ACLINT, with non-backwards-compatible DTS. 

> Are there cases where people really need that specific CLINT RTL instead of a compatible ACLINT?

It depends on what software support for ACLINT looks like, and if users want to run legacy software that does not support ACLINT.

Is there any problem with both `CanHavePeripheryCLINT` and `CanHavePeripheryACLINT`, and disallow `p(CLINTKey).isDefined && p(ACLINTKey).isDefined`

**Comment by ZenithalHourlyRate** [comment]
> Is the spec frozen? How is software support for ACLINT?

It is not frozen, and unfortunately I do not know why. As for software support, [OpenSBI has merged CLINT support to ACLINT](https://github.com/riscv-software-src/opensbi/blob/master/lib/utils/timer/fdt_timer_mtimer.c)

> What is the justification for replacing CLINT with ACLINT in default designs?

ACLINT can be made fully compatible with CLINT, and that can be a default config for ACLINT.

> It depends on what software support for ACLINT looks like, and if users want to run legacy software that does not support ACLINT.

legacy code can use ACLINT just like CLINT for the above default config.

> Is there any problem with both CanHavePeripheryCLINT and CanHavePeripheryACLINT, and disallow p(CLINTKey).isDefined && p(ACLINTKey).isDefined

Technically no.

**Comment by jerryz123** [comment]
Ok. If backwards compatibility can be done, then we can just preserve the old CanHavePeripheryCLINT and CLINTKey, and add flags to CLINTParams to enable non backwards compatible features.

**Review by jerryz123 — CHANGES_REQUESTED** [review]
Can we stick to 2-space indentation in all files?

**Comment by PorterLu** [comment]
Yes, I will make modifications according to the discussion results.

**Comment by PorterLu** [comment]
I have encountered a problem. In order to achieve compatibility with CLINT for ACLINT, I placed the MTIMER device address in ACLINT at 0x0204000. However, in compatibility mode, the size here is 0x7ff8 (mtimecmp size)+8 (mtime size), and AddressSet (0x0204000, 0x8000-1) cannot be successfully compiled.
https://github.com/chipsalliance/rocket-chip/blob/25e2c63567689ebe1fc5e60fdfe3375a8dba071c/src/main/scala/diplomacy/Parameters.scala#L125-L132
Here is my error message:
```
 Caused by: java.lang.IllegalArgumentException: requirement failed: Mis-aligned AddressSets are forbidden, got: AddressSet(0x2004000, 0x7fff)
```

**Comment by PorterLu** [comment]
Related issue: Implement modularization of aclint, which divide clint into mtimer and mswi.

Type of change: feature request

Impact: no functional change

Development Phase: implementation

Release Notes: 
  This PR implements the modular features in ACLINT, dividing CLINT into MTIMER and MSWI. Users can use CLINTKey to generate an ACLINT. When isCompatibleWithClint is true, ACLINT will be compatible with CLINT, otherwise it will not be compatible.
  To address the alignment issue with AddressSet, I set mswi to 0x1ffc000 and mtimer to 0x2000000.

File: src/main/scala/devices/tilelink/CLINT.scala [hunk 1]
```diff
@@ -12,108 +12,46 @@ import freechips.rocketchip.subsystem._
...
 
-    /* 0000 msip hart 0
-     * 0004 msip hart 1
-     * 4000 mtimecmp hart 0 lo
-     * 4004 mtimecmp hart 0 hi
-     * 4008 mtimecmp hart 1 lo
-     * 400c mtimecmp hart 1 hi
-     * bff8 mtime lo
-     * bffc mtime hi
-     */
```
Comment (jerryz123) [inline src/main/scala/devices/tilelink/CLINT.scala:90]: Preserving this comment somewhere would be very helpful.. with ACLINT, it becomes a bit more difficult to understand the address map

Comment (PorterLu) [reply src/main/scala/devices/tilelink/CLINT.scala:90]: Ok, I will write some comments.

**Comment by jerryz123** [comment]
What does the DTS look like when using CLINT compatibility mode?

**Comment by PorterLu** [comment]
Generated Address Map
	       0 -     1000 ARWX  debug-controller@0
	    3000 -     4000 ARWX  error-device@3000
	   10000 -    20000  R X  rom@10000
	 1ffc000 -  2000000 ARW   mswi@1ffc000
	 2000000 -  2008000 ARW   mtimer@2000000
	 c000000 - 10000000 ARW   interrupt-controller@c000000
	60000000 - 80000000  RWX  mmio-port-axi4@60000000
	80000000 - 90000000  RWXC memory@80000000
Like this.

**Comment by jerryz123** [comment]
That's not the DTS

**Comment by jerryz123** [comment]
DTS is the json-esque thing.

**Comment by PorterLu** [comment]
sorry.
/dts-v1/;

/ {
	#address-cells = <1>;
	#size-cells = <1>;
	compatible = "freechips,rocketchip-unknown-dev";
	model = "freechips,rocketchip-unknown";
	L16: cpus {
		#address-cells = <1>;
		#size-cells = <0>;
		timebase-frequency = <1000000>;
		L4: cpu@0 {
			clock-frequency = <0>;
			compatible = "sifive,rocket0", "riscv";
			d-cache-block-size = <64>;
			d-cache-sets = <64>;
			d-cache-size = <16384>;
			d-tlb-sets = <1>;
			d-tlb-size = <32>;
			device_type = "cpu";
			hardware-exec-breakpoint-count = <1>;
			i-cache-block-size = <64>;
			i-cache-sets = <64>;
			i-cache-size = <16384>;
			i-tlb-sets = <1>;
			i-tlb-size = <32>;
			mmu-type = "riscv,sv39";
			next-level-cache = <&L11>;
			reg = <0x0>;
			riscv,isa = "rv64imafdcZicsr_Zifencei_Zihpm_Xrocket";
			riscv,pmpgranularity = <4>;
			riscv,pmpregions = <8>;
			status = "okay";
			timebase-frequency = <1000000>;
			tlb-split;
			L2: interrupt-controller {
				#interrupt-cells = <1>;
				compatible = "riscv,cpu-intc";
				interrupt-controller;
			};
		};
	};
	L11: memory@80000000 {
		device_type = "memory";
		reg = <0x80000000 0x10000000>;
	};
	L15: soc {
		#address-cells = <1>;
		#size-cells = <1>;
		compatible = "freechips,rocketchip-unknown-soc", "simple-bus";
		ranges;
		L8: debug-controller@0 {
			compatible = "sifive,debug-013", "riscv,debug-013";
			debug-attach = "dmi";
			interrupts-extended = <&L2 65535>;
			reg = <0x0 0x1000>;
			reg-names = "control";
		};
		L1: error-device@3000 {
			compatible = "sifive,error0";
			reg = <0x3000 0x1000>;
		};
		L10: external-interrupts {
			interrupt-parent = <&L5>;
			interrupts = <1 2>;
		};
		L5: interrupt-controller@c000000 {
			#interrupt-cells = <1>;
			compatible = "riscv,plic0";
			interrupt-controller;
			interrupts-extended = <&L2 11 &L2 9>;
			reg = <0xc000000 0x4000000>;
			reg-names = "control";
			riscv,max-priority = <3>;
			riscv,ndev = <2>;
		};
		L12: mmio-port-axi4@60000000 {
			#address-cells = <1>;
			#size-cells = <1>;
			compatible = "simple-bus";
			ranges = <0x60000000 0x60000000 0x20000000>;
		};
		L7: mswi@1ffc000 {
			compatible = "riscv,aclint-mswi";
			interrupts-extended = <&L2 3>;
			reg = <0x1ffc000 0x4000>;
			reg-names = "control";
		};
		L6: mtimer@2000000 {
			compatible = "riscv,aclint-mtimer";
			interrupts-extended = <&L2 7>;
			reg = <0x2000000 0x8000>;
			reg-names = "control";
		};
		L13: rom@10000 {
			compatible = "sifive,rom0";
			reg = <0x10000 0x10000>;
			reg-names = "mem";
		};
		L0: subsystem_pbus_clock {
			#clock-cells = <0>;
			clock-frequency = <100000000>;
			clock-output-names = "subsystem_pbus_clock";
			compatible = "fixed-clock";
		};
	};
};

**Comment by jerryz123** [comment]
In normal Clint mode, can we also make it generate the normal CLINT DTS?

I grow more convinced that we should just have separate CanHavePeripheryCLINT and CanHavePeripheryACLINT

**Comment by PorterLu** [comment]
In normal Clint mode, dts must grow like this because of the alignment requirements of the AddressSet. 

> I grow more convinced that we should just have separate CanHavePeripheryCLINT and CanHavePeripheryACLINT

If that's the case, I need to write different code in HasRTC and HasTile based on ACLINT or CLINT, and I need to further investigate this part.

**Comment by PorterLu** [comment]
Related issue: Implement modularization of aclint, which divide clint into mtimer and mswi.

Type of change: feature request

Impact: no functional change

Development Phase: implementation

Release Notes:

1. This PR implements the modular features in ACLINT, dividing CLINT into MTIMER and MSWI. Users can use CLINTKey to generate an ACLINT. When isACLINT is true, ACLINT will be generated, otherwise clint will be generated.

2. If user decide to use clint, the clint base address will be the base address of MSWI device, which is just like normal clint.

3. The reason why I still only use CanHavePeripheryCLINT is that using separate CanHavePeripheryCLINT and CanHavePeripheryACLINT require a lot of code modifications.

Here is clint dts:
```
/ {
	#address-cells = <1>;
	#size-cells = <1>;
	compatible = "freechips,rocketchip-unknown-dev";
	model = "freechips,rocketchip-unknown";
	L15: cpus {
		#address-cells = <1>;
		#size-cells = <0>;
		timebase-frequency = <1000000>;
		L4: cpu@0 {
			clock-frequency = <0>;
			compatible = "sifive,rocket0", "riscv";
			d-cache-block-size = <64>;
			d-cache-sets = <64>;
			d-cache-size = <16384>;
			d-tlb-sets = <1>;
			d-tlb-size = <32>;
			device_type = "cpu";
			hardware-exec-breakpoint-count = <1>;
			i-cache-block-size = <64>;
			i-cache-sets = <64>;
			i-cache-size = <16384>;
			i-tlb-sets = <1>;
			i-tlb-size = <32>;
			mmu-type = "riscv,sv39";
			next-level-cache = <&L10>;
			reg = <0x0>;
			riscv,isa = "rv64imafdcZicsr_Zifencei_Zihpm_Xrocket";
			riscv,pmpgranularity = <4>;
			riscv,pmpregions = <8>;
			status = "okay";
			timebase-frequency = <1000000>;
			tlb-split;
			L2: interrupt-controller {
				#interrupt-cells = <1>;
				compatible = "riscv,cpu-intc";
				interrupt-controller;
			};
		};
	};
	L10: memory@80000000 {
		device_type = "memory";
		reg = <0x80000000 0x10000000>;
	};
	L14: soc {
		#address-cells = <1>;
		#size-cells = <1>;
		compatible = "freechips,rocketchip-unknown-soc", "simple-bus";
		ranges;
		L6: clint@2000000 {
			compatible = "riscv,clint0";
			interrupts-extended = <&L2 3 &L2 7>;
			reg = <0x2000000 0x10000>;
			reg-names = "control";
		};
		L7: debug-controller@0 {
			compatible = "sifive,debug-013", "riscv,debug-013";
			debug-attach = "dmi";
			interrupts-extended = <&L2 65535>;
			reg = <0x0 0x1000>;
			reg-names = "control";
		};
		L1: error-device@3000 {
			compatible = "sifive,error0";
			reg = <0x3000 0x1000>;
		};
		L9: external-interrupts {
			interrupt-parent = <&L5>;
			interrupts = <1 2>;
		};
		L5: interrupt-controller@c000000 {
			#interrupt-cells = <1>;
			compatible = "riscv,plic0";
			interrupt-controller;
			interrupts-extended = <&L2 11 &L2 9>;
			reg = <0xc000000 0x4000000>;
			reg-names = "control";
			riscv,max-priority = <3>;
			riscv,ndev = <2>;
		};
		L11: mmio-port-axi4@60000000 {
			#address-cells = <1>;
			#size-cells = <1>;
			compatible = "simple-bus";
			ranges = <0x60000000 0x60000000 0x20000000>;
		};
		L12: rom@10000 {
			compatible = "sifive,rom0";
			reg = <0x10000 0x10000>;
			reg-names = "mem";
		};
		L0: subsystem_pbus_clock {
			#clock-cells = <0>;
			clock-frequency = <100000000>;
			clock-output-names = "subsystem_pbus_clock";
			compatible = "fixed-clock";
		};
	};
};
```

**Comment by PorterLu** [comment]
Here is aclint dts:
```
/dts-v1/;

/ {
	#address-cells = <1>;
	#size-cells = <1>;
	compatible = "freechips,rocketchip-unknown-dev";
	model = "freechips,rocketchip-unknown";
	L16: cpus {
		#address-cells = <1>;
		#size-cells = <0>;
		timebase-frequency = <1000000>;
		L4: cpu@0 {
			clock-frequency = <0>;
			compatible = "sifive,rocket0", "riscv";
			d-cache-block-size = <64>;
			d-cache-sets = <64>;
			d-cache-size = <16384>;
			d-tlb-sets = <1>;
			d-tlb-size = <32>;
			device_type = "cpu";
			hardware-exec-breakpoint-count = <1>;
			i-cache-block-size = <64>;
			i-cache-sets = <64>;
			i-cache-size = <16384>;
			i-tlb-sets = <1>;
			i-tlb-size = <32>;
			mmu-type = "riscv,sv39";
			next-level-cache = <&L11>;
			reg = <0x0>;
			riscv,isa = "rv64imafdcZicsr_Zifencei_Zihpm_Xrocket";
			riscv,pmpgranularity = <4>;
			riscv,pmpregions = <8>;
			status = "okay";
			timebase-frequency = <1000000>;
			tlb-split;
			L2: interrupt-controller {
				#interrupt-cells = <1>;
				compatible = "riscv,cpu-intc";
				interrupt-controller;
			};
		};
	};
	L11: memory@80000000 {
		device_type = "memory";
		reg = <0x80000000 0x10000000>;
	};
	L15: soc {
		#address-cells = <1>;
		#size-cells = <1>;
		compatible = "freechips,rocketchip-unknown-soc", "simple-bus";
		ranges;
		L8: debug-controller@0 {
			compatible = "sifive,debug-013", "riscv,debug-013";
			debug-attach = "dmi";
			interrupts-extended = <&L2 65535>;
			reg = <0x0 0x1000>;
			reg-names = "control";
		};
		L1: error-device@3000 {
			compatible = "sifive,error0";
			reg = <0x3000 0x1000>;
		};
		L10: external-interrupts {
			interrupt-parent = <&L5>;
			interrupts = <1 2>;
		};
		L5: interrupt-controller@c000000 {
			#interrupt-cells = <1>;
			compatible = "riscv,plic0";
			interrupt-controller;
			interrupts-extended = <&L2 11 &L2 9>;
			reg = <0xc000000 0x4000000>;
			reg-names = "control";
			riscv,max-priority = <3>;
			riscv,ndev = <2>;
		};
		L12: mmio-port-axi4@60000000 {
			#address-cells = <1>;
			#size-cells = <1>;
			compatible = "simple-bus";
			ranges = <0x60000000 0x60000000 0x20000000>;
		};
		L6: mswi@2000000 {
			compatible = "riscv,aclint-mswi";
			interrupts-extended = <&L2 3>;
			reg = <0x2000000 0x4000>;
			reg-names = "control";
		};
		L7: mtimer@2008000 {
			compatible = "riscv,aclint-mtimer";
			interrupts-extended = <&L2 7>;
			reg = <0x2008000 0x8100>;
			reg-names = "control";
		};
		L13: rom@10000 {
			compatible = "sifive,rom0";
			reg = <0x10000 0x10000>;
			reg-names = "mem";
		};
		L0: subsystem_pbus_clock {
			#clock-cells = <0>;
			clock-frequency = <100000000>;
			clock-output-names = "subsystem_pbus_clock";
			compatible = "fixed-clock";
		};
	};
};
```

**Comment by jerryz123** [comment]
Thanks for making this change, the DTS for both cases look fine for me. I'll need to review more thoroughly when I have time, but I think fundamentally this change is OK

File: src/main/scala/devices/tilelink/CLINT.scala [hunk 2]
```diff
@@ -12,108 +12,53 @@ import freechips.rocketchip.subsystem._
...
-          RegField.bytes(t, Some(RegFieldDesc(s"mtimecmp_$i", "", reset=None))))},
-      timeOffset       -> RegFieldGroup("mtime", Some("Timer Register"),
-        RegField.bytes(time, Some(RegFieldDesc("mtime", "", reset=Some(0), volatile=true))))
-    )
-  }
-}
+case object CLINTKey extends Field[Option[CLINTParams]](None)
 
-/** Trait that will connect a CLINT to a subsystem */
+// if is isACLINT is false, a clint will be generated according to base address of mswi device
```
Comment (jerryz123) [inline src/main/scala/devices/tilelink/CLINT.scala:24]:
```suggestion
// if isACLINT is false, a clint will be generated according to base address of mswi device
// The Chisel MSWI device will implement the entirety of the CLINT in this case
```

File: src/main/scala/devices/tilelink/CLINT.scala [hunk 3]
```diff
@@ -12,108 +12,53 @@ import freechips.rocketchip.subsystem._
...
-case object CLINTKey extends Field[Option[CLINTParams]](None)
-
-case class CLINTAttachParams(
-  slaveWhere: TLBusWrapperLocation = CBUS
+// if isACLINT is false, the code will generate a clint
+case class CLINTParams(
+  isACLINT: Boolean = false,
+  mtimer: Option[MTIMERParams]  = Some(MTIMERParams()),
+  mswi: Option[MSWIParams]      = Some(MSWIParams())
 )
```
Comment (jerryz123) [inline src/main/scala/devices/tilelink/CLINT.scala:20]: Does `mtimer=None` or `mswi=None` make sense?

File: src/main/scala/devices/tilelink/CLINT.scala [hunk 4]
```diff
@@ -12,108 +12,53 @@ import freechips.rocketchip.subsystem._
...
-    clint.node := tlbus.coupleTo("clint") { TLFragmenter(tlbus) := _ }
-
-    // Override the implicit clock and reset -- could instead include a clockNode in the clint, and make it a RawModuleImp?
-    InModuleBody {
-      clint.module.clock := tlbus.module.clock
-      clint.module.reset := tlbus.module.reset
+  val mswiOpt = p(CLINTKey) map { clintParams =>
+    val mswiOpt = clintParams.mswi.map { params =>
+      val tlbus = locateTLBusWrapper(p(MSWIAttachKey).slaveWhere)
+      val mswi = LazyModule(new MSWI(params, MTIMERParams(), clintParams.isACLINT, cbus.beatBytes))
```
Comment (jerryz123) [inline src/main/scala/devices/tilelink/CLINT.scala:29]: shouldn't this be clintParams.mtimer?

Comment (jerryz123) [reply src/main/scala/devices/tilelink/CLINT.scala:29]: Also tlbus.beatBytes

File: src/main/scala/subsystem/HasTiles.scala [hunk 5]
```diff
@@ -296,10 +296,22 @@ trait CanAttachTile {
     // 2. The CLINT and PLIC output interrupts are synchronous to the TileLink bus clock,
     //    so might need to be synchronized depending on the Tile's crossing type.
 
-    //    From CLINT: "msip" and "mtip"
-    domain.crossIntIn(crossingParams.crossingType) :=
-      context.clintOpt.map { _.intnode }
-        .getOrElse { NullIntSource(sources = CLINTConsts.ints) }
+    //    From ACLINT: "msip" and "mtip"
+    val clintParams = p(CLINTKey).map { params =>
+      if(params.isACLINT) {
```
Comment (jerryz123) [inline src/main/scala/subsystem/HasTiles.scala:301]:
```suggestion
      if (params.isACLINT) {
```

File: src/main/scala/subsystem/RTC.scala [hunk 6]
```diff
@@ -22,8 +22,15 @@ trait HasRTCModuleImp extends LazyModuleImp {
   // Use the static period to toggle the RTC
   chisel3.withClockAndReset(pbus.module.clock, pbus.module.reset) {
     val (_, int_rtc_tick) = Counter(true.B, internalPeriod.toInt)
-    outer.clintOpt.foreach { clint =>
-      clint.module.io.rtcTick := int_rtc_tick
+    val isACLINT = p(CLINTKey).get.isACLINT
```
Comment (jerryz123) [inline src/main/scala/subsystem/RTC.scala:25]: This needs to support no-clint designs....

**Review by jerryz123 — CHANGES_REQUESTED** [review]

File: src/main/scala/devices/tilelink/CLINT.scala [hunk 7]
```diff
@@ -12,108 +12,53 @@ import freechips.rocketchip.subsystem._
...
+          mtimer.mtimecmpNode := tlbus.coupleTo("mtimecmp") { TLFragmenter(tlbus) := _ }
+          mtimer.mtimeNode := tlbus.coupleTo("mtime") { TLFragmenter(tlbus) := _ }
+
+          InModuleBody {
+            mtimer.module.clock := tlbus.module.clock
+            mtimer.module.reset := tlbus.module.reset
+          }
+
+            mtimer
+          }
```
[the remark this answers is not in the document]
Comment (PorterLu) [reply src/main/scala/devices/tilelink/CLINT.scala:59]: I will fix it

File: src/main/scala/devices/tilelink/CLINT.scala (hunk 4 above)
Comment (PorterLu) [reply src/main/scala/devices/tilelink/CLINT.scala:30]: ClintParams.mtimer is indeed better. But I think it's `cbus. beatBytes`, as clint is cbus device which is consistent with the previous version.

File: src/main/scala/devices/tilelink/CLINT.scala (hunk 3 above)
Comment (PorterLu) [reply src/main/scala/devices/tilelink/CLINT.scala:20]: Yes, let me display dts without clint later.

File: src/main/scala/devices/tilelink/Mtimer.scala [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  def mtimecmpOffset(hart: Int) = hart * mtimecmpBytes
+  def mtimecmpBytes   = 8
+  def mtimeWidth      = 64
+  def mtimecmpSize    = 0x8000
+  def mtimeSize       = 0x100
+  def ints            = 1
+}
+
+// Notice: Remember to ensure that the size and address meet the requirements of the AddressSet, if you use ACLINT 
+case class MTIMERParams(MTIMECMPBaseAddress: BigInt = 0x02008000, MTIMEBaseAddress: BigInt = 0x02010000, intStages: Int = 0)
```
[the remark this answers is not in the document]
Comment (PorterLu) [reply src/main/scala/devices/tilelink/Mtimer.scala]: Thanks. I will fix it.

File: src/main/scala/subsystem/RTC.scala (hunk 6 above)
Comment (PorterLu) [reply src/main/scala/subsystem/RTC.scala:25]: I am making modifications

**Comment by PorterLu** [comment]
Here is no-clint dts:
```
/dts-v1/;

/ {
	#address-cells = <1>;
	#size-cells = <1>;
	compatible = "freechips,rocketchip-unknown-dev";
	model = "freechips,rocketchip-unknown";
	L14: cpus {
		#address-cells = <1>;
		#size-cells = <0>;
		timebase-frequency = <1000000>;
		L4: cpu@0 {
			clock-frequency = <0>;
			compatible = "sifive,rocket0", "riscv";
			d-cache-block-size = <64>;
			d-cache-sets = <64>;
			d-cache-size = <16384>;
			d-tlb-sets = <1>;
			d-tlb-size = <32>;
			device_type = "cpu";
			hardware-exec-breakpoint-count = <1>;
			i-cache-block-size = <64>;
			i-cache-sets = <64>;
			i-cache-size = <16384>;
			i-tlb-sets = <1>;
			i-tlb-size = <32>;
			mmu-type = "riscv,sv39";
			next-level-cache = <&L9>;
			reg = <0x0>;
			riscv,isa = "rv64imafdcZicsr_Zifencei_Zihpm_Xrocket";
			riscv,pmpgranularity = <4>;
			riscv,pmpregions = <8>;
			status = "okay";
			timebase-frequency = <1000000>;
			tlb-split;
			L2: interrupt-controller {
				#interrupt-cells = <1>;
				compatible = "riscv,cpu-intc";
				interrupt-controller;
			};
		};
	};
	L9: memory@80000000 {
		device_type = "memory";
		reg = <0x80000000 0x10000000>;
	};
	L13: soc {
		#address-cells = <1>;
		#size-cells = <1>;
		compatible = "freechips,rocketchip-unknown-soc", "simple-bus";
		ranges;
		L6: debug-controller@0 {
			compatible = "sifive,debug-013", "riscv,debug-013";
			debug-attach = "dmi";
			interrupts-extended = <&L2 65535>;
			reg = <0x0 0x1000>;
			reg-names = "control";
		};
		L1: error-device@3000 {
			compatible = "sifive,error0";
			reg = <0x3000 0x1000>;
		};
		L8: external-interrupts {
			interrupt-parent = <&L5>;
			interrupts = <1 2>;
		};
		L5: interrupt-controller@c000000 {
			#interrupt-cells = <1>;
			compatible = "riscv,plic0";
			interrupt-controller;
			interrupts-extended = <&L2 11 &L2 9>;
			reg = <0xc000000 0x4000000>;
			reg-names = "control";
			riscv,max-priority = <3>;
			riscv,ndev = <2>;
		};
		L10: mmio-port-axi4@60000000 {
			#address-cells = <1>;
			#size-cells = <1>;
			compatible = "simple-bus";
			ranges = <0x60000000 0x60000000 0x20000000>;
		};
		L11: rom@10000 {
			compatible = "sifive,rom0";
			reg = <0x10000 0x10000>;
			reg-names = "mem";
		};
		L0: subsystem_pbus_clock {
			#clock-cells = <0>;
			clock-frequency = <100000000>;
			clock-output-names = "subsystem_pbus_clock";
			compatible = "fixed-clock";
		};
	};
};
```

File: src/main/scala/devices/tilelink/CLINT.scala (hunk 3 above)
Comment (jerryz123) [reply src/main/scala/devices/tilelink/CLINT.scala:20]: In that case, CLINTKey should just be None. I'm not sure I understand why these sub-params need to be option typed too

File: src/main/scala/devices/tilelink/CLINT.scala (hunk 4 above)
Comment (jerryz123) [reply src/main/scala/devices/tilelink/CLINT.scala]: The attach point of the clint is parameterized to the `slaveWhere` of `MSWIAttachKey`, so it should use that bus

File: src/main/scala/devices/tilelink/CLINT.scala (hunk 3 above)
Comment (PorterLu) [reply src/main/scala/devices/tilelink/CLINT.scala:20]: Because in the spec,  The mtimer may not exist or the mswi may not exist.

**Comment by jerryz123** [comment]
Please squash the github-reply-to-suggestion commits into one commit.
FYI, there is also an option to batch apply those changes

File: src/main/scala/devices/tilelink/CLINT.scala (hunk 4 above)
Comment (PorterLu) [reply src/main/scala/devices/tilelink/CLINT.scala]: That's true, but I'm consistent with the previous version of CLINT. I think sequencer believes that CLINT will only be available on CBUS.

File: src/main/scala/devices/tilelink/CLINT.scala (hunk 3 above)
Comment (jerryz123) [reply src/main/scala/devices/tilelink/CLINT.scala:20]: Ok. Can you add a require to this class to check that both cannot be set to None?

Also, can you verify that all three cases generate properly?

Comment (PorterLu) [reply src/main/scala/devices/tilelink/CLINT.scala:20]: Yes, I will check three cases:
```
mtimer: not None, mswi: not None
mtimer: None, mswi: not None
mtimer: not None, mswi: None
```

**Comment by PorterLu** [comment]
1. I set beatBytes according bus type.
2. I do not allow mtimerParams and mswiParams to be empty at same time.

File: src/main/scala/devices/tilelink/Mtimer.scala [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  slaveWhere: TLBusWrapperLocation = CBUS
+)
+
+case object MTIMERAttachKey extends Field(MTIMERAttachParams())
+
+class MTIMER(params: MTIMERParams, beatBytes: Int)(implicit p: Parameters) extends LazyModule
+{
+  import MTIMERConsts._ 
+
+  val device = new SimpleDevice("mtimer", Seq("riscv,aclint-mtimer")) {
```
Comment (ZenithalHourlyRate) [inline src/main/scala/devices/tilelink/Mtimer.scala:43]: mtimer exposes `reg = <0x2008000 0x8100>;` to the dts, which does not show where mtimer base address is. We might see how opensbi find mtimer base address

https://github.com/riscv-software-src/opensbi/blob/dc1c7db05e075e0910b93504370b50d064a51402/lib/utils/timer/fdt_timer_mtimer.c#L71-L74

Seems like it should be something like `reg = <0x2008000 0x7FF0 0x2010000 0x8>;`

File: src/main/scala/devices/tilelink/MSWI.scala [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+      Annotated.params(this, mtimerParams)
+    }
+
+    require (intnode.edges.in.size == 0, "MSWI only produces interrupts; it does not accept them")
+
+    val nTiles  = intnode.out.size
+    val ipi     = Seq.fill(nTiles) { RegInit(0.U(1.W)) }
+    
+    val io = IO(new Bundle {
+        val rtcTick = Input(Bool())
```
Comment (ZenithalHourlyRate) [inline src/main/scala/devices/tilelink/MSWI.scala:90]: It should be an option.

File: src/main/scala/devices/tilelink/Mtimer.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (ZenithalHourlyRate) [reply src/main/scala/devices/tilelink/Mtimer.scala:26]: This is not addressed

[the remark this answers is not in the document]
Comment (ZenithalHourlyRate) [reply src/main/scala/devices/tilelink/Mtimer.scala:26]: Seems impossible due to the requirement of address alignment

**Review by ZenithalHourlyRate — CHANGES_REQUESTED** [review]

File: src/main/scala/devices/tilelink/MSWI.scala [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+     * 0000 msip hart 0
+     * 0004 msip hart 1
+     * 4000 mtimecmp hart 0 lo
+     * 4004 mtimecmp hart 0 hi
+     * 4008 mtimecmp hart 1 lo
+     * 400c mtimecmp hart 1 hi
+     * bff8 mtime lo
+     * bffc mtime hi
+     */
+    val mtimerRegMapping = if(!isACLINT) Seq(
```
Comment (ZenithalHourlyRate) [inline src/main/scala/devices/tilelink/MSWI.scala:115]:
```suggestion
    val mtimerRegMapping = if (!isACLINT) Seq(
```

Space between if and bracket

File: src/main/scala/devices/tilelink/Mtimer.scala [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+     *
+     * 0 mtime lo
+     * 4 mtime hi
+     */
+    mtimecmpNode.regmap(0 -> timecmpRegGroup)
+    mtimeNode.regmap(0 -> timeRegGroup)
+  }
+}
+
+object MTIMER {
```
Comment (ZenithalHourlyRate) [inline src/main/scala/devices/tilelink/Mtimer.scala:91]: Since it is proved to be convenient, you might have a `object SWI` for the current MSWI and your future SSWI impl

**Review by ZenithalHourlyRate — COMMENTED** [review]
I am fairly happy with this implementation now. Comments left should still be addressed. I need some time to test it personally to see if it works, and I also need some time to investigate into the diplomacy issue proposed in private communication.

**Review by ZenithalHourlyRate — COMMENTED** [review]
I am fairly happy with this implementation now. Comments left should still be addressed. I need some time to test it personally to see if it works, and I also need some time to investigate into the diplomacy issue proposed in private communication.

**Comment by ZenithalHourlyRate** [comment]
@PorterLu, @cyyself, and we together have managed to boot Linux with OpenSBI and a patched u-boot (thanks to @lbmeng) on a dual-core Rocket with ACLINT of this branch enabled on a 7k325t FPGA. I think this suffices for testing.

As for the dts issue, due to diplomacy, what rocket will generate for aclint mtimer is
```
reg = <0x2010000 0x1000 0x2008000 0x8000>;
```
instead of the example in the RFC spec (https://lore.kernel.org/lkml/20211007123632.697666-9-anup.patel@wdc.com/)
```
reg = <0x2010000 0x8>, <0x2008000 0x7ff8>;
```
though in practice OpenSBI and the patched u-boot worked happily.

This is ready for more review. And for my part, after some minor fixes (e.g. code style, a WithACLINT option, and a config for emulator in CI), I think this can be merged.

**Comment by ZenithalHourlyRate** [comment]
> As for the dts issue, due to diplomacy, what rocket will generate for aclint mtimer is
> 
> ```
> reg = <0x2010000 0x1000 0x2008000 0x8000>;
> ```
> 
> instead of the example in the RFC spec (https://lore.kernel.org/lkml/20211007123632.697666-9-anup.patel@wdc.com/)
> 
> ```
> reg = <0x2010000 0x8>, <0x2008000 0x7ff8>;
> ```
> 
> though in practice OpenSBI and the patched u-boot worked happily.

dtc is not too happy

```
Warning (simple_bus_reg): /soc/mtimer@2008000: simple-bus unit address format error, expected "2010000"
```

for

```
        L7: mtimer@2008000 {
            compatible = "riscv,aclint-mtimer";
            interrupts-extended = <&L2 7>;
            reg = <0x2010000 0x1000 0x2008000 0x8000>;
            reg-names = "control";
        };
```

changing the node name to `mtimer@2010000` would eliminate the warning. But the example in the RFC dt-binding does use `mtimer@2004000` instead of `mtimer@200bff8`.
